File: python/src/starline/wire_pack.py
"""8-byte bitpacked perceptual state — Phase IV compression."""

from __future__ import annotations

import struct
from typing import Sequence

PACKED_SIZE = 8
V_MAX = 200.0
TWO_PI = 6.283185307179586


def _clamp(x: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, x))


def _q10(x: float) -> int:
    return int(round(_clamp(x, 0.0, 1.0) * 1023))


def _q12_unit(x: float, scale: float) -> int:
    return int(round(_clamp(x / scale, 0.0, 1.0) * 4095))


def pack_perceptual(
    *,
    v: float,
    smoothness: float,
    coherence: float,
    phase: float,
    sync_R: float,
    flow_code: int = 0,
    topology_code: int = 15,
    v_max: float = V_MAX,
) -> bytes:
    """Pack perceptual core into 8 bytes (little-endian uint64 subset)."""
    ci = _q10(coherence)
    si = _q10(smoothness)
    vi = _q12_unit(v, v_max)
    phii = _q12_unit(phase % TWO_PI, TWO_PI)
    ri = _q10(sync_R)
    fi = int(flow_code) & 3
    ti = int(topology_code) & 15

    val = (
        ci
        | (si << 10)
        | (vi << 20)
        | (phii << 32)
        | (ri << 44)
        | (fi << 54)
        | (ti << 56)
    )
    return struct.pack("<Q", val & 0xFFFFFFFFFFFFFFFF)


def unpack_perceptual(data: bytes, *, v_max: float = V_MAX) -> dict[str, float | int]:
    if len(data) < PACKED_SIZE:
        data = data + b"\x00" * (PACKED_SIZE - len(data))
    (val,) = struct.unpack("<Q", data[:PACKED_SIZE])
    ci = val & 0x3FF
    si = (val >> 10) & 0x3FF
    vi = (val >> 20) & 0xFFF
    phii = (val >> 32) & 0xFFF
    ri = (val >> 44) & 0x3FF
    fi = (val >> 54) & 0x3
    ti = (val >> 56) & 0xF
    return {
        "coherence": ci / 1023.0,
        "smoothness": si / 1023.0,
        "v": vi / 4095.0 * v_max,
        "phase": phii / 4095.0 * TWO_PI,
        "sync_R": ri / 1023.0,
        "flow_code": fi,
        "topology_code": ti,
    }


def pack_from_wire(wire: Sequence[float], *, v_max: float = V_MAX) -> bytes:
    """Pack from full wire[20] layout (docs/ENCODING.md indices 1-based in Lua)."""
    w = list(wire)
    if len(w) < 18:
        w = w + [0.0] * (18 - len(w))
    return pack_perceptual(
        v=float(w[6]),
        smoothness=float(w[3]),
        coherence=float(w[10]),
        phase=float(w[13]),
        sync_R=float(w[15]),
        flow_code=int(w[17]),
        topology_code=int(w[5]) if w[5] < 255 else 15,
        v_max=v_max,
    )


def wire_from_packed(data: bytes, template_wire: Sequence[float] | None = None) -> list[float]:
    """Expand packed core into wire[20], preserving G slots from template if given."""
    p = unpack_perceptual(data)
    w = list(template_wire) if template_wire else [0.0] * 20
    while len(w) < 20:
        w.append(0.0)
    w[3] = float(p["smoothness"])
    w[5] = float(p["topology_code"])
    w[6] = float(p["v"])
    w[10] = float(p["coherence"])
    w[13] = float(p["phase"])
    w[15] = float(p["sync_R"])
    w[17] = float(p["flow_code"])
    return w

File: roblox/Starline/WirePack.lua
--!strict
--[[ 8-byte perceptual pack — mirror python/starline/wire_pack.py ]]

local WirePack = {}

WirePack.PACKED_SIZE = 8
WirePack.V_MAX = 200

local function clamp(x: number, lo: number, hi: number): number
	return math.max(lo, math.min(hi, x))
end

local function q10(x: number): number
	return math.floor(clamp(x, 0, 1) * 1023 + 0.5)
end

local function q12unit(x: number, scale: number): number
	return math.floor(clamp(x / scale, 0, 1) * 4095 + 0.5)
end

function WirePack.encodeFromWire(wire: { number }): buffer
	local v = wire[7] or 0
	local S = wire[4] or 0
	local c = wire[11] or 0
	local phi = wire[14] or 0
	local R = wire[16] or 0
	local flow = wire[18] or 0
	local topo = wire[6] or 15
	if topo >= 255 then
		topo = 15
	end

	local ci = q10(c)
	local si = q10(S)
	local vi = q12unit(v, WirePack.V_MAX)
	local phii = q12unit(phi % (math.pi * 2), math.pi * 2)
	local ri = q10(R)
	local fi = math.floor(flow) % 4
	local ti = math.floor(topo) % 16

	local val = ci + si * 2 ^ 10 + vi * 2 ^ 20 + phii * 2 ^ 32 + ri * 2 ^ 44 + fi * 2 ^ 54 + ti * 2 ^ 56

	local buf = buffer.create(WirePack.PACKED_SIZE)
	buffer.writeu32(buf, 0, val % 2 ^ 32)
	buffer.writeu32(buf, 4, math.floor(val / 2 ^ 32))
	return buf
end

function WirePack.decodeToWire(packed: buffer, template: { number }?): { number }
	local w = template or table.create(20, 0)
	for i = 1, 20 do
		if w[i] == nil then
			w[i] = 0
		end
	end
	local lo = buffer.readu32(packed, 0)
	local hi = buffer.readu32(packed, 4)
	local val = lo + hi * 2 ^ 32
	local ci = val % 2 ^ 10
	local si = math.floor(val / 2 ^ 10) % 2 ^ 10
	local vi = math.floor(val / 2 ^ 20) % 2 ^ 12
	local phii = math.floor(val / 2 ^ 32) % 2 ^ 12
	local ri = math.floor(val / 2 ^ 44) % 2 ^ 10
	local fi = math.floor(val / 2 ^ 54) % 4
	local ti = math.floor(val / 2 ^ 56) % 16

	w[4] = si / 1023
	w[6] = ti
	w[7] = vi / 4095 * WirePack.V_MAX
	w[11] = ci / 1023
	w[14] = phii / 4095 * (math.pi * 2)
	w[16] = ri / 1023
	w[18] = fi
	return w
end

return WirePack
